File: valid-parentheses/shaun2.cpp
//shaun's own idea, first sub wrong answer failing case "(])"
//modified after for this failed case, accepted
//runtime 4ms, faster than 93%, mem less than 69%
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (char ch: s) {
            if (ch == '(' || ch == '{' || ch == '[') st.push(ch);
            else {
                if (st.empty()) return false;
                //can match
                if ((ch == ')' && st.top() == '(') || 
                    (ch == ']' && st.top() == '[') || 
                    (ch == '}' && st.top() == '{')) st.pop();
                //else st.push(ch); // first fix
                else return false; //okay too
            }
        }
        return st.empty();
    }
};

File: valid-parentheses/shaun2.py
#Runtime: 20 ms, faster than 97.82% of Python3 online submissions for Valid Parentheses.
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '[' or c == '{':
                stack.append(c)
            else:
                if len(stack) == 0: return False
                t = stack.pop()
                if  t == '(' and c != ')' \
                or t =='[' and c != ']' \
                or t == '{' and c != '}':
                    return False
        return len(stack) == 0
        
